README.rst
*******************************************************
``autoprop`` --- Infer properties from accessor methods
*******************************************************
.. image:: https://img.shields.io/pypi/v/autoprop.svg
   :alt: Last release
   :target: https://pypi.python.org/pypi/autoprop

.. image:: https://img.shields.io/pypi/pyversions/autoprop.svg
   :alt: Python version
   :target: https://pypi.python.org/pypi/autoprop

.. image:: 
   https://img.shields.io/github/workflow/status/kalekundert/autoprop/Test%20and%20release/master
   :alt: Test status
   :target: https://github.com/kalekundert/autoprop/actions

.. image:: https://img.shields.io/coveralls/kalekundert/autoprop.svg
   :alt: Test coverage
   :target: https://coveralls.io/github/kalekundert/autoprop?branch=master

.. image:: https://img.shields.io/github/last-commit/kalekundert/autoprop?logo=github
   :alt: GitHub last commit
   :target: https://github.com/kalekundert/autoprop

``autoprop`` is a library for automatically filling in classes with properties 
(e.g. ``obj.x``) corresponding to each accessor method (e.g. ``obj.get_x()``, 
``obj.set_x()``).  The biggest reasons to use ``autoprop`` are:

- Less boilerplate than defining properties manually.

- Sophisticated support for cached properties.

Installation
============
Install ``autoprop`` using ``pip``::

    $ pip install autoprop

Usage
=====
To use ``autoprop``, import the ``autoprop`` module and use it directly as a 
class decorator::

    >>> import autoprop
    >>>
    >>> @autoprop
    ... class Vector2D(object):
    ...    
    ...     def __init__(self, x, y):
    ...         self._x = x
    ...         self._y = y
    ...
    ...     def get_x(self):
    ...         return self._x
    ...
    ...     def set_x(self, x):
    ...         self._x = x
    ...
    ...     def get_y(self):
    ...         return self._y
    ...
    ...     def set_y(self, y):
    ...         self._y = y
    ...
    >>> v = Vector2D(1, 2)
    >>> v.x, v.y
    (1, 2)

The decorator searches your class for methods beginning with ``get_``, 
``set_``, or ``del_`` and uses them to create properties.  The names of the 
properties are taken from whatever comes after the underscore.  For example, 
the method ``get_x`` would be used to make a property called ``x``.  Any 
combination of getter, setter, and deleter methods is allowed for each 
property.

Caching
=======
If you have properties that are expensive to calculate, it's easy to cache 
them::

    >>> @autoprop.cache
    ... class Simulation(object):
    ...
    ...     def get_data(self):
    ...         print("expensive calculation...")
    ...         return 42
    ...
    >>> s = Simulation()
    >>> s.data
    expensive calculation...
    42
    >>> s.data
    42

It's also easy to cache some properties but not others::

    >>> @autoprop
    ... class Simulation(object):
    ...
    ...     def get_cheap(self):
    ...         print("cheap calculation...")
    ...         return 16
    ...
    ...     @autoprop.cache
    ...     def get_expensive(self):
    ...         print("expensive calculation...")
    ...         return 42
    ...
    >>> s = Simulation()
    >>> s.cheap
    cheap calculation...
    16
    >>> s.cheap
    cheap calculation...
    16
    >>> s.expensive
    expensive calculation...
    42
    >>> s.expensive
    42

The ``@autoprop.cache()`` decorator accepts a ``policy`` keyword argument that 
determines how the cache will be managed.  The following policies are 
supported:

- ``overwrite``: This is the default policy.  Values are cached by overwriting 
  the property itself, such that future lookups will directly access the cached 
  value with no overhead.  This is exactly equivalent to using 
  `functools.cached_property`.  Unlike normal properties, there is no way to 
  customize what happens when setting or deleting these properties.  Setting 
  the property will update its value, and deleting it will cause its value to 
  be recalculated on the next access.

- ``manual``: Cached values are never recalculated automatically, but can be 
  recalculated and/or changed manually.  There are two ways to do this:
  
  1. Specify ``provide_mutators=True`` to ``@autoprop.cache()``.  This will 
     instruct autoprop to provide default setter and deleter implementations 
     for the property, which will allow the cached value to be changed or 
     dropped, respectively.  
    
  2. Call ``autoprop.set_cached_attr()`` and/or ``autoprop.del_cached_attr()``.  
     These functions allow you to implement your own setter and deleter 
     functions, which is often the entire purpose of using this policy.
  
  This policy has ≈10x more overhead than the ``overwrite`` policy, but allows 
  you to control what happens when the attribute is set or deleted (like a 
  regular property).  

- ``automatic``: Cached values are automatically recalculated if certain other 
  attributes of the object change.  In order to use this policy, you must 
  specify ``watch=<list of attributes>`` to ``@autoprop.cache()``.  The *watch* 
  argument must be iterable, and each item must either be the name of an 
  attribute (e.g. a string) or a callable that will accept the object in 
  question and return any value.  The cached value will be recalculated 
  whenever any of the "watched" values change.  The cache can also be 
  recalculated manually, in any of the ways described for the ``manual`` 
  policy.

  This policy has ≈25x more overhead than the ``overwrite`` policy, but allows 
  cached values to stay up to date when the attributes they depend on change.
  
- ``immutable``: Properties are never recalculated, and are furthermore not 
  allowed to have setter or deleter methods (an error will be raised if any 
  such methods are found).  As the name implies, this is for properties and 
  classes that are intended to be immutable.  
  
  Note that ``@autoprop.immutable`` is an alias for 
  ``@autoprop.cache(policy='immutable')``.

- ``dynamic``: Properties are recalculated every time they are accessed.  This 
  is exactly equivalent what ``autoprop`` does when caching is disabled, which 
  is exactly equivalent to using ``@property``.  Use this policy when you want 
  to specify ``@autoprop.cache`` at the class-level, but also need to prevent a 
  few properties from being cached.
  
  Note that ``@autoprop.dynamic`` is an alias for 
  ``@autoprop.cache(policy='dynamic')``.

Details
=======
Besides having the right prefix, there are two other criteria that methods must 
meet in order to be made into properties.  The first is that they must take the 
right number of arguments.  Getters and deleters must not require any arguments 
(other than self).  Setters must accept exactly one argument (other than self), 
which is the value to set.  Default, variable, and keyword arguments are all 
ignored; all that matters is that the function can be called with the expected 
number of arguments.

Any methods that have the right name but the wrong arguments are silently 
ignored.  This can be nice for getters that require, for example, an index.  
Even though such a getter can't be made into a property, ``autoprop`` allows it 
to follow the same naming conventions as any getters that can be::

    >>> @autoprop
    ... class Vector2D(Vector2D):
    ...     
    ...     def get_coord(self, i):
    ...         if i == 0: return self.x
    ...         if i == 1: return self.y
    ...
    ...     def set_coord(self, i, new_coord):
    ...         if i == 0: self.x = new_coord
    ...         if i == 1: self.y = new_coord
    ...
    >>> v = Vector2D(1, 2)
    >>> v.get_x()
    1
    >>> v.get_coord(0)
    1

In this way, users of your class can always expect to find accessors named 
``get_*`` and ``set_*``, and properties corresponding to those accessors for 
basic attributes that don't need any extra information.

The second criterion is that the property must have a name which is not already 
in use.  This guarantees that nothing you explicitly add to your class will be 
overwritten, and it gives you the ability to manually customize how certain 
properties are defined if you'd so like.  This criterion does not apply to 
superclasses, so it is possible for properties to shadow attributes defined in 
parent classes.

If you want to explicitly ignore a method which would otherwise be discovered 
by ``autoprop``, use the ``@autoprop.ignore`` decorator.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v4.1.0 (2022-05-30)
### Feature
* Add decorator to explicitly ignore methods ([`1687e06`](https://github.com/kalekundert/autoprop/commit/1687e062421f76b6d497297779a3e7f3b30bd52d))

### Fix
* Handle undefined getters in error messages ([`f2bb9e5`](https://github.com/kalekundert/autoprop/commit/f2bb9e5b7586d5e09f3cc14690b6e15c1ab6863e))

## v4.0.2 (2021-09-15)
### Fix
* Upgrade to signature_dispatch~=1.0 ([`453447a`](https://github.com/kalekundert/autoprop/commit/453447aecdd7ddc7ca9e7898d0f6b99ff4a073f0))

## v4.0.1 (2021-09-10)
### Fix
* Use built-in functools.cached_property in py3.8+ ([`be358a9`](https://github.com/kalekundert/autoprop/commit/be358a9cf4b95d60a70ecbe318852f9db63c92d1))

## v4.0.0 (2021-09-09)
### Feature
* Add overwrite/automatic/manual cache policies ([`a013b69`](https://github.com/kalekundert/autoprop/commit/a013b69e1222918e776899d81eab29417eab0af0))

### Breaking
* remove property/object/class cache policies.  ([`a013b69`](https://github.com/kalekundert/autoprop/commit/a013b69e1222918e776899d81eab29417eab0af0))

### Documentation
* Describe how to manually clear the cache ([`7a526e7`](https://github.com/kalekundert/autoprop/commit/7a526e73b9c9ca39d9813cdeb10abba8d1f73059))

## v3.0.0 (2021-05-27)
### Feature
* Change default cache policy to 'property' ([`a3baec5`](https://github.com/kalekundert/autoprop/commit/a3baec55ceee721c9090dc2a5639ac686eb6170e))

### Breaking
* the new default caches more aggressively, so code written with the old default may end up with stale values.  Update old code by replacing `@autoprop.cache` with `@autoprop.cache(policy='object')`.  ([`a3baec5`](https://github.com/kalekundert/autoprop/commit/a3baec55ceee721c9090dc2a5639ac686eb6170e))

## v2.2.0 (2021-04-05)
### Feature
* Add the `immutable` decorator ([`e8f023a`](https://github.com/kalekundert/autoprop/commit/e8f023ab099b3d1477bb285267e11528c43c4f5c))

## v2.1.1 (2021-04-01)
### Fix
* Cache getters (not just properties) ([`ab454eb`](https://github.com/kalekundert/autoprop/commit/ab454eb1d247b147675fc6c377747892588762c1))

### Documentation
* Reorganize the caching section ([`f7ac9bb`](https://github.com/kalekundert/autoprop/commit/f7ac9bb79ac7764f9227601dd6108c756e934704))
* Fix typo ([`613e180`](https://github.com/kalekundert/autoprop/commit/613e180067aa646e9696e70b88435f3962b37c5c))
* Tweak wording ([`b009007`](https://github.com/kalekundert/autoprop/commit/b009007a819d5bdb71c82dae7d539dca2804edf9))
* Fix formatting ([`11b027f`](https://github.com/kalekundert/autoprop/commit/11b027f36a71f0d94352d3f6911844d0308f222b))
* Improve the first example ([`8e365a1`](https://github.com/kalekundert/autoprop/commit/8e365a1b19f1ab3553ec9346e6e2ed59f132b029))

## v2.1.0 (2021-03-31)
### Feature
* Only configure caching if requested ([`109f430`](https://github.com/kalekundert/autoprop/commit/109f430fd286e139ed2fef02777d106be2334df2))

### Fix
* Don't have *every* static/class method invalidate the cache ([`a231052`](https://github.com/kalekundert/autoprop/commit/a231052fb038260055ad5bdaa860e2e25eb4f5a3))

### Documentation
* Document the caching system ([`1974c42`](https://github.com/kalekundert/autoprop/commit/1974c42e0f33859c6c282869cbb861b2c9a8a3de))

## v2.0.0 (2021-03-30)
### Feature
* Drop support for python 3.5 ([`d4781cb`](https://github.com/kalekundert/autoprop/commit/d4781cb489a0a74987ca96eb390d4e1dd507ff47))
* Implement caching policies ([`a05726a`](https://github.com/kalekundert/autoprop/commit/a05726aa96ccce670274430180272dd2c72c3df5))

### Breaking
* Drop support for python2.  ([`a05726a`](https://github.com/kalekundert/autoprop/commit/a05726aa96ccce670274430180272dd2c72c3df5))

### Documentation
* Remove python2 badge ([`815c60d`](https://github.com/kalekundert/autoprop/commit/815c60d5d81d82f567baba2df4b927bc4cbbc224))


tests/conftest.py
#!/usr/bin/env python3

import sys

def pytest_ignore_collect(path, config):
    if sys.version_info[0] == 2:
        return 'python3' in str(path)
    if sys.version_info[0] == 3:
        return 'python2' in str(path)


tests/test_python2.py
#!/usr/bin/env python2

import pytest
import autoprop

def test_complain_about_old_style_classes():
    with pytest.raises(TypeError) as err:
        @autoprop
        class Example:
            pass

    assert '@autoprop can only be used with new-style classes' in str(err)



autoprop/cache.py
_CACHE_ATTR = '__autoprop_cache'
_SET_BY_USER = object()
_UNSPECIFIED = object()

class Cache:

    def __init__(self, obj):
        self.values = {}
        self.memos = {}

class CachedProperty(property):

    def __set_name__(self, owner, name):
        self.name = name

    def __get__(self, obj, owner=None):
        # Class attribute access (e.g. for docstrings): 
        if obj is None:
            return super().__get__(obj, owner)

        # Instance attribute access:
        cache = get_cache(obj)
        attr = self.name

        try:
            return cache.values[attr]
        except KeyError:
            value = super().__get__(obj, owner)
            cache.values[attr] = value
            return value

class ConditionalCachedProperty(property):

    def __init__(self, getter, setter, deleter, memo_manager):
        super().__init__(getter, setter, deleter)
        self.memo_manager = memo_manager

    def __set_name__(self, owner, name):
        self.name = name

    def __get__(self, obj, owner=None):
        # Class attribute access (e.g. for docstrings): 
        if obj is None:
            return super().__get__(obj, owner)

        # Instance attribute access:
        cache = get_cache(obj)
        attr = self.name
        manager = self.memo_manager
        curr_memo = manager.refresh(obj)

        try:
            prev_memo = cache.memos[attr]
        except KeyError:
            is_fresh = False
        else:
            is_fresh = prev_memo is _SET_BY_USER or \
                    manager.is_fresh(prev_memo, curr_memo)

        if is_fresh:
            # Assume that a `memos` entry implies a `values` entry.
            return cache.values[attr]

        value = super().__get__(obj, owner)
        cache.values[attr] = value
        cache.memos[attr] = curr_memo
        return value

def get_cache(obj):
    try:
        return getattr(obj, _CACHE_ATTR)
    except AttributeError:
        cache = Cache(obj)
        setattr(obj, _CACHE_ATTR, cache)
        return cache

def get_cached_attr(obj, attr, default=_UNSPECIFIED):
    """
    Get the cached value of the given attribute.

    .. warning::

        This method should only be used on properties that are cached with the 

    Arguments:
        obj:
            Any object.

        attr (str):
            The name of the attribute to look up.

        default:
            The value to return if the given attribute is not found in the 
            cache.  If not specified, an :exc:`AttributeError` will be raised 
            in this circumstance.

    If the given object didn't have a cache, this will initialize and empty one.
    Only the 
    """
    cache = get_cache(obj)
    try:
        return cache.values[attr]
    except KeyError as err:
        if default is not _UNSPECIFIED:
            return default
        else:
            raise AttributeError(err) from None

def set_cached_attr(obj, attr, value):
    cache = get_cache(obj)
    cache.values[attr] = value
    cache.memos[attr] = _SET_BY_USER

def del_cached_attr(obj, attr):
    cache = get_cache(obj)

    try:
        del cache.values[attr]
    except KeyError as err:
        raise AttributeError(err) from None

    try:
        del cache.memos[attr]
    except KeyError:
        pass

def clear_cache(obj):
    """
    Delete the cache associated with the given object.

    .. warning::

        Only the ``manual``, ``automatic``, and ``immutable`` cache policies 
        will be affected by this operation.  The ``overwrite`` policy caches 
        values directly in the instance dictionary, and so will not be 
        affected.

    This will force any values stored in the cache to be recalculated the next 
    time they are needed.
    """
    try:
        delattr(obj, _CACHE_ATTR)
    except AttributeError:
        pass




autoprop/__init__.py
#!/usr/bin/env python3

"""
Infer properties from accessor methods.

The :mod:`autoprop` module can be directly used as a class decorator.  

- Method discovery
  - Private methods

- Argument signatures
  - optional argument allowed
  - required arguments must match
"""

from .decorators import (
        autoprop, cache, dynamic, immutable, policy, ignore
)
from .cache import (
        get_cache, clear_cache,
        get_cached_attr, set_cached_attr, del_cached_attr,
)

__version__ = '4.1.0'

# Hack to make the module directly usable as a decorator.  Only works for 
# python 3.5 or higher.  See this Stack Overflow post:
# https://stackoverflow.com/questions/1060796/callable-modules

import sys

class _CallableModule(sys.modules[__name__].__class__):

    def __call__(self, *args, **kwargs):
        return autoprop(*args, **kwargs)

sys.modules[__name__].__class__ = _CallableModule
del sys, _CallableModule



autoprop.py
#!/usr/bin/env python

"""Infer properties from accessor methods."""

import sys
import inspect
import re
import functools
import signature_dispatch
from functools import partial
from collections import defaultdict

__version__ = '3.0.0'

_EXPECTED_NUM_ARGS = {'get': 0, 'set': 1, 'del': 0}
_CACHE_POLICIES = ['dynamic', 'class', 'object', 'property', 'immutable']
_CACHE_ATTR = '_autoprop_cache'
_CACHE_VERSION_ATTR = '_autoprop_cache_version'
_CACHE_POLICY_ATTR = '_autoprop_cache_policy'
_REFRESH_ATTR = '_autoprop_refresh'
_INITIAL_VERSION = 1
_DEFAULT_POLICY = 'property'

class Cache:

    def __init__(self, obj):
        self.obj = obj
        self.cls = obj.__class__

        self.values = {}
        self.cls_versions = {}
        self.obj_versions = {}
        self.attr_touched = {}

    def bind(self, attr):
        return BoundCache(self, attr)

class BoundCache:

    def __init__(self, cache, attr):
        self.cache = cache
        self.attr = attr

    def is_stale(self, policy):
        if policy == 'dynamic':
            return True

        cache = self.cache
        attr = self.attr

        if attr not in cache.values:
            return True

        if policy == 'immutable':
            return False

        if cache.attr_touched.get(attr, False):
            return True

        if policy == 'property':
            return False

        v0 = _INITIAL_VERSION - 1

        if cache.obj_versions.get(attr, v0) < get_cache_version(cache.obj):
            return True

        if policy == 'object':
            return False

        assert policy == 'class'

        for base in cache.cls.__mro__:
            my_version = cache.cls_versions.get(base, {}).get(attr, v0)
            base_version = get_cache_version(base)
            if my_version < base_version:
                return True

        return False

    def get(self):
        return self.cache.values[self.attr]

    def update(self, value):
        cache = self.cache
        attr = self.attr

        cache.values[attr] = value
        cache.attr_touched[attr] = False
        cache.obj_versions[attr] = get_cache_version(cache.obj)

        # Build this dictionary on the fly to be robust against the (remote) 
        # possibility that `__mro__` could change at runtime, e.g.:
        # https://stackoverflow.com/questions/20822850/change-python-mro-at-runtime 

        for base in cache.cls.__mro__:
            cache.cls_versions.setdefault(base, {})[attr] = \
                    get_cache_version(base)

    def touch(self):
        self.cache.attr_touched[self.attr] = True

class CachedProperty(property):

    def __init__(self, name, getter, setter, deleter, policy):
        super().__init__(getter, setter, deleter)
        self.name = name
        self.policy = policy

    def __get__(self, obj, owner=None):
        # Class attribute access (e.g. for docstrings): 
        if obj is None:
            return super().__get__(obj, owner)

        # Instance attribute access:
        cache = self.get_bound_cache(obj)

        if cache.is_stale(self.policy):
            value = super().__get__(obj, owner)
            cache.update(value)

        return cache.get()

    def __set__(self, obj, value):
        cache = self.get_bound_cache(obj)
        cache.touch()
        return super().__set__(obj, value)

    def __delete__(self, obj):
        cache = self.get_bound_cache(obj)
        cache.touch()
        return super().__delete__(obj)

    def get_bound_cache(self, obj):
        cache = get_cache(obj)
        return cache.bind(self.name)

    def getter_replacement(self):
        def getter(obj, *args, **kwargs):
            if not args and not kwargs:
                return self.__get__(obj, None)
            else:
                return self.fget(obj, *args, **kwargs)
        return getter

def autoprop(cls):
    return _make_autoprops(cls)

def cache(*args, **kwargs):

    def wrapper(x, policy):
        if inspect.isclass(x):
            return _make_autoprops(x, cache=True, default_policy=policy)
        else:
            return _assign_policy(x, policy=policy)

    dispatch = signature_dispatch()

    @dispatch
    def decorator(*, policy):
        _check_policy(policy)
        return partial(wrapper, policy=policy)

    @dispatch
    def decorator():
        return partial(wrapper, policy=_DEFAULT_POLICY)

    @dispatch
    def decorator(f):
        return wrapper(f, _DEFAULT_POLICY)

    return decorator(*args, **kwargs)

def dynamic(f):
    return cache(policy='dynamic')(f)

def immutable(f):
    return cache(policy='immutable')(f)

def policy(policy):
    _check_policy(policy)

    def wrapper(f):
        _assign_policy(f, policy)
        return f

    return wrapper

def refresh(f):
    setattr(f, _REFRESH_ATTR, True)
    return f

def get_cache(obj):
    try:
        cache = getattr(obj, _CACHE_ATTR)
    except AttributeError:
        cache = Cache(obj)
        setattr(obj, _CACHE_ATTR, cache)

    return cache

def get_cache_version(obj):
    return getattr(obj, _CACHE_VERSION_ATTR, _INITIAL_VERSION)

def clear_cache(obj):
    try:
        delattr(obj, _CACHE_ATTR)
    except AttributeError:
        pass


def _make_autoprops(cls, *, cache=False, default_policy=_DEFAULT_POLICY):
    accessors = defaultdict(dict)

    def get_accessor(name, prefix):
        # If we found the accessor in this class, return it.
        try:
            return accessors[name][prefix]

        # Otherwise, look for a suitable method in parent classes.
        except KeyError:

            try:
                full_name = '%s_%s' % (prefix, name)
                attr = getattr(cls, full_name)
            except AttributeError:
                return None
            else:
                return attr if _is_accessor(full_name, attr) else None

    def increment_cache_version(f, predicate=lambda *args, **kwargs: True):
        if f is None:
            return f

        @functools.wraps(f)
        def wrapper(*args, **kwargs):
            if predicate(*args, **kwargs):
                if args and isinstance(args[0], cls):
                    obj = args[0]
                else:
                    obj = cls

                version = get_cache_version(obj)
                setattr(obj, _CACHE_VERSION_ATTR, version + 1)

            return f(*args, **kwargs)

        return wrapper

    for name, attr in cls.__dict__.items():
        # Because we are iterating through `__dict__`, class/static methods 
        # will not be bound and will not appear as functions, which is good 
        # because we cannot make class properties without metaclasses.

        key = _is_accessor(name, attr)
        if key:
            name, prefix = key
            accessors[name][prefix] = attr

            if hasattr(attr, _CACHE_POLICY_ATTR):
                if not cache:
                    raise ValueError("\n".join([
                        f"cache disabled: can't specify cache policies",
                        f"class: {cls.__qualname__}",
                        f"method: {attr.__qualname__}",
                        f"decorate the class with `@autoprop.cache` or `@autoprop.dynamic` to enable the cache.",
                    ]))


        if getattr(attr, _REFRESH_ATTR, False):
            if not cache:
                raise ValueError("\n".join([
                    f"cache disabled: can't use `@autoprop.refresh`",
                    f"class: {cls.__qualname__}",
                    f"method: {attr.__qualname__}",
                    f"decorate the class with `@autoprop.cache` or `@autoprop.dynamic` to enable the cache.",
                ]))
            setattr(cls, name, increment_cache_version(attr))

        # We have to dig inside static/class methods to see if the underlying 
        # function is supposed to get decorated.

        if isinstance(attr, (classmethod, staticmethod)):
            if getattr(attr.__func__, _REFRESH_ATTR, False):
                if not cache:
                    raise ValueError("\n".join([
                        f"cache disabled: can't use `@autoprop.refresh`",
                        f"class: {cls.__qualname__}",
                        f"method: {attr.__func__.__qualname__}",
                        f"decorate the class with `@autoprop.cache` or `@autoprop.dynamic` to enable the cache.",
                    ]))
                attr = type(attr)(increment_cache_version(attr.__func__))
                setattr(cls, name, attr)

    for name in accessors:

        # Don't overwrite any attributes defined in this class.  Attributes 
        # defined in superclasses may be shadowed.

        if name in cls.__dict__:
            continue

        getter  = get_accessor(name, 'get')
        setter  = get_accessor(name, 'set')
        deleter = get_accessor(name, 'del')

        if not cache:
            # Mark the getter as 'dynamic'.  This won't have any affect on this 
            # class, but if a subclass with caching enabled incorporates this 
            # getter into an autoprop, this policy will ensure that the getter 
            # remains un-cached (as its author presumably intended).
            if getter:
                dynamic(getter)

            prop = property(getter, setter, deleter)

        else:
            setter = increment_cache_version(setter)
            deleter = increment_cache_version(deleter)
            policy = getattr(getter, _CACHE_POLICY_ATTR, default_policy)

            if policy == 'immutable':
                if getter and setter:
                    raise ValueError("\n".join([
                        f"can't specify setter for immutable property",
                        f"property: {cls.__qualname__}.{name}",
                        f"immutable getter: {getter.__qualname__}",
                        f"setter: {setter.__qualname__}",
                    ]))
                if getter and deleter:
                    raise ValueError("\n".join([
                        f"can't specify deleter for immutable property",
                        f"property: {cls.__qualname__}.{name}",
                        f"immutable getter: {getter.__qualname__}",
                        f"deleter: {deleter.__qualname__}",
                    ]))

            prop = CachedProperty(name, getter, setter, deleter, policy)

            if getter:
                setattr(cls, f'get_{name}', prop.getter_replacement())

        setattr(cls, name, prop)

    if cache:

        # Automatically increment the cache on attribute assignment.  
        # 
        # Don't do the same for attribute deletion, because (i) classes are not 
        # supposed to implement `__delattr__()` unless they actually support it 
        # and (ii) it's easy to get this behavior by decorating `__delattr__()` 
        # with `@autoprop.refresh`.  Decorating `__setattr__()` is not so 
        # easy because steps must be taken to avoid infinite recursion.

        try:
            __setattr__ = cls.__dict__['__setattr__']
        except KeyError:
            def __setattr__(self, name, value):
                super(cls, self).__setattr__(name, value)

        def predicate(self, name, value):
            return not name.startswith('_autoprop')

        __setattr__ = increment_cache_version(__setattr__, predicate)
        setattr(cls, '__setattr__', __setattr__)

    return cls

def _check_policy(policy):
    if policy not in _CACHE_POLICIES:
        expected = ', '.join(repr(p) for p in _CACHE_POLICIES)
        raise ValueError(f"unknown policy {policy!r}, expected one of: {expected}")

def _assign_policy(f, policy):
    if not f.__name__.startswith('get_'):
        raise ValueError(f"can't cache {f.__qualname__}; it's not a getter")

    setattr(f, _CACHE_POLICY_ATTR, policy)
    return f

def _is_accessor(name, attr):
    if not inspect.isfunction(attr):
        return False

    accessor_match = re.match('(get|set|del)_(.+)', name)
    if not accessor_match:
        return False

    def is_required(p):
        positional = p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD
        return p.kind in positional and p.default is p.empty

    # The attributes passed to this function must be accessed via `getattr()` 
    # (e.g. as opposed to `cls.__dict__[]`) because static/class methods are 
    # not callable until bound to the class.

    prefix, name = accessor_match.groups()
    sig = inspect.signature(attr)
    params = sig.parameters.values()
    num_args = sum(map(is_required, params)) - 1

    if num_args != _EXPECTED_NUM_ARGS[prefix]:
        return False

    return name, prefix

# Hack to make the module directly usable as a decorator.  Only works for 
# python 3.5 or higher.  See this Stack Overflow post:
# https://stackoverflow.com/questions/1060796/callable-modules

class CallableModule(sys.modules[__name__].__class__):

    def __call__(self, *args, **kwargs):
        return autoprop(*args, **kwargs)

sys.modules[__name__].__class__ = CallableModule

